### PR TITLE
Opset12 upgrade for existing models used by perf/e2e pipelines

### PR DESCRIPTION
**Pull request by ytaous**
**Description**: This is part of opset 12 upgrade work.
The current bert/gpt-2 models used for perf test and e2e pipelines are at opset 9/10. Those models were  previously transformed thus containing the "TrainableDropout" node. Since in Opset 12, we want to support Dropout-12 spec, thus we need to write a custom scripts to replace those TrainableDropout with Dropout in the models.    

**Motivation and Context**
A few tests were done locally:
- new util scripts to transform models
- fixed an issue in large gpt-2 model
- copy new models to dev0 for pipeline tests
- create new blob and upload it to Azuyre Cloud for download
- perf test and e2e run locally
- new calibrated csv
- log running convergence comparison

### CODE REVIEW

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+#replace TrainableDropout with Dropout
+process_trainabledropout(model)
+# some gpt-2 models (large ones) still don't have this input corrected
+align_attention_mask_dim(model)
+
+#set opset version to 12
+model.opset_import[0].version = 12
+
+f = open(output_model_name, "wb")
```
Comment (thiagocrepaldi) [inline orttraining/tools/scripts/opset12_model_transform.py:128]:
```python
with  open (output_model_name, "wb") as f:
    f.write(model.SerializeToString())
```
is safer and a best practice for managing files

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+f.write(model.SerializeToString())
+f.close()
+
+#
+# To verify the converted model in case of bert, refer to the code at the end of model_transform.py 
+#
```
Comment (thiagocrepaldi) [inline orttraining/tools/scripts/opset12_model_transform.py:134]: NIT: missing new line at the end

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    if t_value is not None:
+        attr.type = 4
+        attr.t.CopyFrom(t_value)
+    else:
+        attr.type = 1
+        attr.f = f_value
+    return const_node
+
+def process_trainabledropout(model):
+    delete_nodels = []
```
Comment (thiagocrepaldi) [inline orttraining/tools/scripts/opset12_model_transform.py:77]: NIT: is 'nodels' an intended name or a typo for 'nodes'?

**Comment by ytaous** [comment]
I ran training 640 steps on both bert-L and large gpt-2, the total runtime is almost the same before and after upgrade.

---
In reply to: [430886864](https://github.com/microsoft/onnxruntime/pull/4238#issuecomment-430886864) [](ancestors = 430886864)

File: orttraining/tools/ci_test/download_e2e_test_data.py [hunk 4]
```diff
@@ -11,8 +11,8 @@
 import zipfile
 
 # update these if the E2E test data changes
-ARCHIVE_BLOB_URL = "https://onnxruntimetestdata.blob.core.windows.net/training/onnxruntime_training_data.zip?snapshot=2020-04-14T02:10:05.3158245Z"
-ARCHIVE_SHA256_DIGEST = "ea4168a801ded478f4e2af08232cb1174913caac300d5bf73b2652dc6894372c"
+ARCHIVE_BLOB_URL = "https://onnxruntimetestdata.blob.core.windows.net/training/onnxruntime_training_data_v12.zip?snapshot=2020-06-13T06:24:15.0833240Z"
```
Comment (edgchen1) [inline orttraining/tools/ci_test/download_e2e_test_data.py:14]: >onnxruntime_training_data_v12 [](start = 79, length = 29)

do we need a new one or can we use the same archive name?

Comment (ytaous) [reply orttraining/tools/ci_test/download_e2e_test_data.py:14]: I prefer new one is there's concern, we can deprecate old one later

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+def find_input_node(model, arg):
+    result = []
+    for node in model.graph.node:
+        for output in node.output:
+            if output == arg:
+                result.append(node)
+    return result[0] if len(result)== 1 else None
+
+def find_output_node(model, arg):
```
Comment (edgchen1) [inline orttraining/tools/scripts/opset12_model_transform.py:39]: >find_output_node [](start = 4, length = 16)

is this used?

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+    result = []
+    for node in model.graph.node:
+        for input in node.input:
+            if input == arg:
+                result.append(node)
+    return result[0] if len(result) == 1 else None
+
+def find_input(model, arg):
```
Comment (edgchen1) [inline orttraining/tools/scripts/opset12_model_transform.py:47]: >find_input [](start = 4, length = 10)

is this used?

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    for initializer in model.graph.initializer:
+        if initializer.name == arg:
+            return initializer
+    return None
+
+def get_node_index(model, node):
+    i = 0
```
Comment (edgchen1) [inline orttraining/tools/scripts/opset12_model_transform.py:54]: >i = 0 [](start = 4, length = 5)

fyi, the enumerate function might be useful here. it provides an index as well.
[https://docs.python.org/3/library/functions.html#enumerate](https://docs.python.org/3/library/functions.html#enumerate)

File: orttraining/tools/scripts/opset12_model_transform.py (hunk 6 above)
Comment (ytaous) [reply orttraining/tools/scripts/opset12_model_transform.py:47]: Initial was, forgot to clean it up. will keep in my TODOs and clean it up in next PR.

File: orttraining/tools/scripts/opset12_model_transform.py (hunk 7 above)
Comment (ytaous) [reply orttraining/tools/scripts/opset12_model_transform.py:54]: will do this in my next PR

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+if len(sys.argv) < 2:
+    print("Please give model path...")
+    exit(1)
+
+input_model_name = sys.argv[1]
+output_model_name = input_model_name[:-5] + '_opset12.onnx'
+
+model = onnx.load(input_model_name)
+
+def find_input_node(model, arg):
```
Comment (edgchen1) [inline orttraining/tools/scripts/opset12_model_transform.py:31]: >find_input_node [](start = 4, length = 15)

this function is called find_input_node, but it looks through node outputs. i'm a bit confused by the name.

File: orttraining/tools/scripts/opset12_model_transform.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    for initializer in model.graph.initializer:
+        if initializer.name == arg:
+            return initializer
+    return None
+
+def get_node_index(model, node):
+    i = 0
+    while i < len(model.graph.node):
+        if model.graph.node[i] == node:
+            break;
```
Comment (edgchen1) [inline orttraining/tools/scripts/opset12_model_transform.py:57]: >break; [](start = 12, length = 6)

usually don't need a semicolon in python
also can we just return i here?

File: orttraining/tools/scripts/opset12_model_transform.py (hunk 8 above)
Comment (ytaous) [reply orttraining/tools/scripts/opset12_model_transform.py:31]: Sorry about that, the code was inherited from previous one, I can include the change in my next PR to clarify more.

File: orttraining/tools/ci_test/download_e2e_test_data.py (hunk 4 above)
Comment (edgchen1) [reply orttraining/tools/ci_test/download_e2e_test_data.py:14]: i don't see why we need two currently - the archive is versioned by the snapshots. if we will run two separate sets of tests which require different versions, keeping them separate makes sense.

---
In reply to: [440456503](https://github.com/microsoft/onnxruntime/pull/4238#discussion_r440456503) [](ancestors = 440456503)

Comment (ytaous) [reply orttraining/tools/ci_test/download_e2e_test_data.py:14]: OK, let me work on it
